File: parse/parser.py
from lex import Lexer, Token, Loc, LocPtr, Tok
from semantic_analysis import Scope, ScopeFlags
from utils.diagnostic import diag, Diag

CUR_PARSER = None

class Parser:
    lexer: Lexer
    tok: Token
    prev_tok_location: Loc
    paren_count: int
    brace_count: int
    bracket_count: int
    cur_scope: Scope

    def __init__(self, lexer: Lexer):
        self.lexer = lexer
        self.tok = lexer.lex()
        self.prev_tok_location = 0
        self.paren_count = 0
        self.brace_count = 0
        self.bracket_count = 0
        self.cur_scope = Scope(None, 0)
        global CUR_PARSER
        CUR_PARSER = self

    def unconsume_token(self, consumed: Token):
        next_tok = self.tok
        self.lexer.enter_token(consumed, True)
        self.tok = self.lexer.lex()
        self.lexer.enter_token(next_tok, True)

    def consume_token(self) -> Loc:
        assert not self.is_token_special(), "Should consume special tokens with consume_*_token"
        self.prev_tok_location = self.tok.loc
        self.tok = self.lexer.lex()
        return self.prev_tok_location

    def try_consume_token(self, expected: Tok, loc_ptr: LocPtr = None) -> bool:
        if self.tok.ty != expected:
            return False
        loc = self.consume_token()
        if loc_ptr is not None:
            loc_ptr.value = loc
        return True

    def consume_any_token(self) -> Loc:
        if self.is_token_paren():
            return self.consume_paren()
        if self.is_token_brace():
            return self.consume_brace()
        if self.is_token_bracket():
            return self.consume_bracket()
        return self.consume_token()

    def is_token_paren(self) -> bool:
        return self.tok.ty == Tok.LPAREN or self.tok.ty == Tok.RPAREN

    def is_token_brace(self) -> bool:
        return self.tok.ty == Tok.LBRACE or self.tok.ty == Tok.RBRACE

    def is_token_bracket(self) -> bool:
        return self.tok.ty == Tok.LSQUARE or self.tok.ty == Tok.RSQUARE

    def is_token_special(self) -> bool:
        return self.is_token_paren() or self.is_token_brace() or self.is_token_bracket()

    def consume_paren(self) -> Loc:
        assert self.is_token_paren(), "Wrong consume method"
        if self.tok.ty == Tok.LPAREN:
            self.paren_count += 1
        elif self.paren_count > 0:
            # self.angle_brackets.clear()
            self.paren_count -= 1
        self.prev_tok_location = self.tok.loc
        self.tok = self.lexer.lex()
        return self.prev_tok_location

    def consume_brace(self) -> Loc:
        assert self.is_token_brace(), "Wrong consume method"
        if self.tok.ty == Tok.LBRACE:
            self.brace_count += 1
        elif self.brace_count > 0:
            # self.angle_brackets.clear()
            self.brace_count -= 1
        self.prev_tok_location = self.tok.loc
        self.tok = self.lexer.lex()
        return self.prev_tok_location

    def consume_bracket(self) -> Loc:
        assert self.is_token_bracket(), "Wrong consume method"
        if self.tok.ty == Tok.LSQUARE:
            self.bracket_count += 1
        elif self.bracket_count > 0:
            # self.angle_brackets.clear()
            self.bracket_count -= 1
        self.prev_tok_location = self.tok.loc
        self.tok = self.lexer.lex()
        return self.prev_tok_location

    def expect_and_consume(self, expected: Tok, diag_id: str = "expected {}", msg: str = "") -> bool:
        if self.tok.ty == expected:
            self.consume_any_token()
            return False
        if is_common_typo(expected, self.tok.ty):
            loc = self.tok.loc
            # fixit with replacement ...
            diag(loc, diag_id + " " + msg, Diag.ERROR)
            self.consume_any_token()
            return False
        # fixit ...
        diag_msg = ""
        if diag_id == "expected {}":
            diag_msg = diag_id.format(expected)
        elif diag_id == "expected {} after {}":
            diag_msg = diag_id.format(expected, msg)
        else:
            diag_msg = diag_id.format(msg)
        loc = self.tok.loc # should get loc of end of previous token
        diag(loc, diag_msg, Diag.ERROR)
        return True

    def expect_and_consume_semi(self, diag_id: str = "expected ';'", token_used: str = "") -> bool:
        if self.try_consume_token(Tok.SEMI):
            return False
        if self.tok.ty in [Tok.RPAREN, Tok.RSQUARE] and self.next_token().ty == Tok.SEMI:
            diag(self.tok.loc, f"extraneous '{self.tok.ty}' before ';'", Diag.ERROR) #fixit
            self.consume_any_token()
            self.consume_token()
            return False
        return self.expect_and_consume(Tok.SEMI, diag_id, token_used)

    def skip_until(self, *until_toks, **options):
        toks = list(until_toks)
        stop_before_match = ("stop_before_match" in options.keys()) and options[
            "stop_before_match"
        ]
        stop_at_semi = ("stop_at_semi" in options.keys()) and options["stop_at_semi"]
        is_first_token_skipped = True
        while True:
            for i in range(len(toks)):
                if self.tok.ty == toks[i]:
                    if not stop_before_match:
                        self.consume_any_token()
                    return True
            if len(toks) == 1 and toks[0] == Tok.EOF and not stop_at_semi:
                while self.tok.ty != Tok.EOF:
                    self.consume_any_token()
                return True
            match self.tok.ty:
                case Tok.EOF:
                    return False
                case Tok.LPAREN:
                    self.consume_paren()
                    self.skip_until(Tok.RPAREN)
                case Tok.LSQUARE:
                    self.consume_bracket()
                    self.skip_until(Tok.RSQUARE)
                case Tok.LBRACE:
                    self.consume_brace()
                    self.skip_until(Tok.RBRACE)
                case Tok.QUESTION:
                    self.consume_token()
                    self.skip_until(
                        Tok.COLON,
                        stop_at_semi=True,
                        stop_before_match=stop_before_match,
                    )
                case Tok.RPAREN:
                    if self.paren_count > 0 and not is_first_token_skipped:
                        return False
                    self.consume_paren()
                case Tok.RSQUARE:
                    if self.bracket_count > 0 and not is_first_token_skipped:
                        return False
                    self.consume_bracket()
                case Tok.RBRACE:
                    if self.brace_count > 0 and not is_first_token_skipped:
                        return False
                    self.consume_brace()
                case Tok.SEMI:
                    if stop_at_semi:
                        return False
                    self.consume_token()
                case _:
                    self.consume_any_token()
            is_first_token_skipped = False

    def enter_scope(self, scope_flags: ScopeFlags):
        self.cur_scope = Scope(self.cur_scope, scope_flags)

    def exit_scope(self):
        assert self.cur_scope is not None
        # actions.act_on_pop_scope(self.tok.loc, self.cur_scope)
        old_scope = self.cur_scope
        assert old_scope.parent_scope is not None
        self.cur_scope = old_scope.parent_scope

    def parse(self) -> []:
        from .decl import parse_translation_unit
        return parse_translation_unit()


def is_common_typo(expected: Tok, actual: Tok) -> bool:
    match expected:
        case Tok.SEMI:
            return actual == Tok.COLON or actual == Tok.COMMA
        case _:
            return False

def parser() -> Parser:
    assert CUR_PARSER is not None
    return CUR_PARSER

File: cppver/compiler/lib/parse/parser.cpp

#include "parser.hpp"
#include "ast/nodes/type.hpp"
#include "diags/diagnostic.hpp"
#include "lexer/tok.hpp"
#include "prec.hpp"
#include <cassert>
#include <cstring>

Prec prec_from_tok(Tok kind) {
  switch (kind) {
  case tok::EQUAL:
  case tok::STAREQUAL:
  case tok::SLASHEQUAL:
  case tok::PERCENTEQUAL:
  case tok::PLUSEQUAL:
  case tok::MINUSEQUAL:
  case tok::LESSLESSEQUAL:
  case tok::GREATERGREATEREQUAL:
  case tok::AMPEQUAL:
  case tok::PIPEEQUAL:
  case tok::CARETEQUAL:
    return Prec::ASSIGN;
  case tok::QUESTION:
    return Prec::COND;
  case tok::PIPEPIPE:
    return Prec::OR;
  case tok::AMPAMP:
    return Prec::AND;
  case tok::PIPE:
    return Prec::BOR;
  case tok::CARET:
    return Prec::XOR;
  case tok::AMP:
    return Prec::BAND;
  case tok::EXCLAIMEQUAL:
  case tok::EQUALEQUAL:
    return Prec::EQ;
  case tok::GREATEREQUAL:
  case tok::GREATER:
  case tok::LESSEQUAL:
  case tok::LESS:
    return Prec::COMP;
  case tok::LESSLESS:
  case tok::GREATERGREATER:
    return Prec::SHIFT;
  case tok::PLUS:
  case tok::MINUS:
    return Prec::PLUS;
  case tok::PERCENT:
  case tok::SLASH:
  case tok::STAR:
    return Prec::STAR;
  default:
    return Prec::UNKNOWN;
  }
}

void Parser::unconsume_token(Token consumed) {
  lexer.enter_token(tok, true);
  tok = consumed;
  // auto next_tok = tok;
  // lexer.enter_token(consumed, true);
  // tok = lexer.lex();
  // lexer.enter_token(next_tok, true);
}

Loc Parser::consume_token() {
  assert(!is_token_special());
  prev_tok_location = tok.loc;
  tok = lexer.lex();
  return prev_tok_location;
}

bool Parser::try_consume_token(Tok expected, Loc *loc) {
  if (tok.kind != expected)
    return false;
  auto l = consume_token();
  if (loc)
    *loc = l;
  return true;
}

Loc Parser::consume_any_token() {
  if (is_token_paren())
    return consume_paren();
  if (is_token_square())
    return consume_square();
  if (is_token_brace())
    return consume_brace();
  return consume_token();
}

Loc Parser::consume_paren() {
  assert(is_token_paren());
  if (tok.kind == tok::LPAREN)
    paren_count++;
  else if (paren_count > 0)
    paren_count--;
  prev_tok_location = tok.loc;
  tok = lexer.lex();
  return prev_tok_location;
}

Loc Parser::consume_brace() {
  assert(is_token_brace());
  if (tok.kind == tok::LBRACE)
    brace_count++;
  else if (brace_count > 0)
    brace_count--;
  prev_tok_location = tok.loc;
  tok = lexer.lex();
  return prev_tok_location;
}

Loc Parser::consume_square() {
  assert(is_token_square());
  if (tok.kind == tok::LSQUARE)
    square_count++;
  else if (square_count > 0)
    square_count--;
  prev_tok_location = tok.loc;
  tok = lexer.lex();
  return prev_tok_location;
}

bool Parser::expect_and_consume(Tok expected, char const *diag_id,
                                char const *str) {
  if (tok.kind == expected) {
    consume_any_token();
    return false;
  }
  auto loc = tok.loc;
  if (is_common_typo(expected, tok.kind)) {
    // fixit with replacement...
    Diag(diag::ERROR, loc, "%s %s", diag_id, str);
    consume_any_token();
    return false;
  }
  // fixit
  if (strcmp(diag_id, "expected %s") == 0)
    Diag(diag::ERROR, loc, diag_id, tok::get_name(expected));
  else if (strcmp(diag_id, "expected %s after %s") == 0)
    Diag(diag::ERROR, loc, diag_id, tok::get_name(expected), str);
  else
    Diag(diag::ERROR, loc, diag_id, str);
  return true;
}

bool Parser::expect_and_consume_semi(char const *diag_id,
                                     char const *token_used) {

  if (try_consume_token(tok::SEMI))
    return false;
  // if ((tok.kind == tok::RPAREN || tok.kind == tok::RSQUARE) &&
  //     next_token().kind == tok::SEMI) {
  //   Diag(diag::ERROR, tok.loc, "extraneous '%s' before ';'",
  //        tok::get_name(tok.kind));
  //   consume_any_token();
  //   consume_token();
  //   return false;
  // }
  return expect_and_consume(tok::SEMI, diag_id, token_used);
}

bool Parser::skip_until(std::span<Tok> toks, bool at_semi, bool before_match) {
  bool is_first_token_skipped = true;
  while (true) {
    for (auto t : toks) {
      if (tok.kind == t) {
        if (!before_match)
          consume_any_token();
        return true;
      }
    }
    if (toks.size() == 1 && toks[0] == tok::END_OF_FILE && !at_semi) {
      while (tok.kind != tok::END_OF_FILE)
        consume_any_token();
      return true;
    }
    switch (tok.kind) {
    case tok::END_OF_FILE:
      return false;
    case tok::LPAREN:
      consume_paren();
      skip_until(tok::RPAREN);
      break;
    case tok::LSQUARE:
      consume_square();
      skip_until(tok::RSQUARE);
      break;
    case tok::LBRACE:
      consume_brace();
      skip_until(tok::RBRACE);
      break;
    case tok::QUESTION:
      consume_token();
      skip_until(tok::COLON, true, before_match);
      break;
    case tok::RPAREN:
      if (paren_count > 0 && !is_first_token_skipped)
        return false;
      consume_paren();
      break;
    case tok::RSQUARE:
      if (square_count > 0 && !is_first_token_skipped)
        return false;
      consume_square();
      break;
    case tok::RBRACE:
      if (brace_count > 0 && !is_first_token_skipped)
        return false;
      consume_brace();
      break;
    case tok::SEMI:
      if (at_semi)
        return false;
      consume_token();
      break;
    default:
      consume_any_token();
    }
    is_first_token_skipped = false;
  }
}

bool is_common_typo(Tok expected, Tok actual) {
  if (expected == tok::SEMI)
    return actual == tok::COLON || actual == tok::COMMA;
  return false;
}

// TYPE functions

Type *Parser::parse_type() {
  Type *cur_type = nullptr;
  while (true) {
    if (cur_type == nullptr) {
      if (tok::is_builtin_type(tok.kind)) {
        cur_type = sema.ctx.get_builtin_type_from_tok(tok.kind);
        consume_token();
        continue;
      }
      if (tok.kind == tok::IDENT) {
        auto type_name = tok.ident();
        auto type_loc = consume_token();
        auto lookup_decl = cur_scope->lookup_named_decl(type_name);
        TypeDecl *td = nullptr;
        if (lookup_decl == nullptr ||
            !(td = lookup_decl->dyn_cast<TypeDecl>())) {
          Diag(diag::ERROR, type_loc, "'%s' is not a type name",
               type_name->name.c_str());
          return nullptr;
        }
        cur_type = sema.ctx.get_declared_type(td);
        continue;
      }
      break;
    }
    if (tok.kind == tok::STAR) {
      cur_type = sema.ctx.get_pointer_type(cur_type);
      consume_token();
      continue;
    }
    if (tok.kind == tok::LSQUARE) {
      consume_square();
      auto [count, _] = parse_integer_constexpr();
      expect_and_consume(tok::RSQUARE);
      cur_type = sema.ctx.get_array_type(cur_type, count);
      continue;
    }
    // function type ?
    break;
  }
  return cur_type;
}

bool Parser::is_start_of_type() {
  if (tok::is_builtin_type(tok.kind))
    return true;
  if (tok.kind == tok::IDENT) {
    auto type_name = tok.ident();
    auto lookup_decl = cur_scope->lookup_named_decl(type_name);
    return lookup_decl && lookup_decl->dyn_cast<TypeDecl>();
  }
  return false;
}

bool Parser::following_is_type(int) { return false; }
